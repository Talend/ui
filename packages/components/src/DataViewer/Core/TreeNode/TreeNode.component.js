import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';
import Tree from '../Tree';

/**
 * Return the type of the current node.
 * @param {string} type
 */
export function isBranch(type) {
	return type === 'array' || type === 'object';
}

/**
 * Check if the branch is currently opened with the help of the jonspath.
 * @param {boolean} expandAll
 * @param {number} index
 * @param {array} paths
 * @param {string} jsonpath
 */
export function isBranchOpened(expandAll, index, paths, jsonpath) {
	const isPresent = get(paths, index, []).indexOf(jsonpath) !== -1;
	return expandAll ? !isPresent : isPresent;
}

/**
 * Check if the node is highlight with the help of the jsonpath.
 * @param {array} highlighted
 * @param {string} jsonpath
 */
export function isNodeHighlighted(highlighted, jsonpath) {
	return !!highlighted.find(pattern => jsonpath.match(pattern));
}

/**
 * Check if children's branch has been selected and branch is closed.
 * if so the branch will be highlighted.
 * Only for object type.
 * @param {boolean} opened
 * @param {boolean} nodeHighlighted
 * @param {string} type
 */
export function isDeepNodeHighlighted(opened, nodeHighlighted, type) {
	if (type === 'object' || type === 'array') {
		return !opened && nodeHighlighted;
	}
	return nodeHighlighted;
}

/**
 * Render a branch or a leaf.
 * It gives the recursiveTree func to the branch to help render another tree.
 */
export default class TreeNode extends React.Component {
	static propTypes = {
		branch: PropTypes.func.isRequired,
		getItemType: PropTypes.func.isRequired,
		index: PropTypes.number,
		isAllExpanded: PropTypes.bool,
		jsonpath: PropTypes.string.isRequired,
		leaf: PropTypes.func.isRequired,
		level: PropTypes.number,
		paths: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
		value: PropTypes.any,
	};

	recursiveTree = args => <Tree {...this.props} {...args} level={this.props.level + 1} />;

	render() {
		const type = this.props.getItemType(this.props.value);
		const opened = isBranchOpened(
			this.props.isAllExpanded,
			this.props.index,
			this.props.paths,
			this.props.jsonpath,
		);
		const nodeHighlighted = isNodeHighlighted(
			get(this.props, 'highlighted', []),
			this.props.jsonpath,
		);

		if (isBranch(type)) {
			return this.props.branch({
				...this.props,
				nodeHighlighted: isDeepNodeHighlighted(opened, nodeHighlighted, type),
				opened,
				recursive: this.recursiveTree,
				type,
			});
		}
		return this.props.leaf({ ...this.props, nodeHighlighted });
	}
}
