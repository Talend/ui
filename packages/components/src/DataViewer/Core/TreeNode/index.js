import Component from './TreeNode.component';

export default Component;
