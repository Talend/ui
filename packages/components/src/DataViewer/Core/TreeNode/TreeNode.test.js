import React from 'react';
import { shallow } from 'enzyme';
import TreeNode, {
	isBranch,
	isBranchOpened,
	isNodeHighlighted,
	isDeepNodeHighlighted,
} from './TreeNode.component';

describe('isBranch', () => {
	it('should return true', () => {
		expect(isBranch('array')).toEqual(true);
	});
	it('should return true', () => {
		expect(isBranch('object')).toEqual(true);
	});
	it('should return false', () => {
		expect(isBranch('something')).toEqual(false);
	});
});

describe('isBranchOpened', () => {
	it('should return true with expandAll false', () => {
		const paths = ['jsonpath'];
		const ret = isBranchOpened(false, 0, paths, 'jsonpath');
		expect(ret).toEqual(true);
	});
	it('should return false with expandAll false', () => {
		const paths = [];
		const ret = isBranchOpened(false, 0, paths, 'jsonpath');
		expect(ret).toEqual(false);
	});
	it('should return false with expandAll true', () => {
		const paths = ['jsonpath'];
		const ret = isBranchOpened(true, 0, paths, 'jsonpath');
		expect(ret).toEqual(false);
	});
	it('should return true with expandAll true', () => {
		const paths = [];
		const ret = isBranchOpened(true, 0, paths, 'jsonpath');
		expect(ret).toEqual(true);
	});
});

describe('isNodeHighlighted', () => {
	it('should return true, the node is highlighted', () => {
		expect(isNodeHighlighted(['jsonpath'], 'jsonpath')).toEqual(true);
	});
	it('should return false, the node is not highlighted', () => {
		expect(isNodeHighlighted([''], 'jsonpath')).toEqual(false);
	});
});

describe('isDeepNodeHighlighted', () => {
	it('shoul return true, when its an object and its not opened', () => {
		expect(isDeepNodeHighlighted(false, true, 'ojbect')).toBe(true);
	});
	it('shoul return true, when its something else than object and highlist is true', () => {
		expect(isDeepNodeHighlighted(false, true, 'something')).toBe(true);
	});
	it('shoul return false, when highlist is false', () => {
		expect(isDeepNodeHighlighted(false, false, 'object')).toBe(false);
	});
});

describe('TreeNode', () => {
	it('should return a new branch', () => {
		const props = {
			level: 0,
			value: {},
			jsonpath: '',
			index: 0,
		};
		const branch = jest.fn();
		branch.mockReturnValueOnce(<div>new branch</div>);
		const leaf = jest.fn();
		const getItemType = jest.fn();
		getItemType.mockReturnValueOnce('array');
		const wrapper = shallow(
			<TreeNode {...props} getItemType={getItemType} branch={branch} leaf={leaf} />,
		);
		expect(branch).toHaveBeenCalled();
		expect(wrapper.getElement()).toMatchSnapshot();
	});
	it('should return a new leaf', () => {
		const props = {
			level: 0,
			value: {},
			jsonpath: '',
			index: 0,
		};
		const branch = jest.fn();
		const leaf = jest.fn();
		leaf.mockReturnValueOnce(<div>new leaf</div>);
		const getItemType = jest.fn();
		getItemType.mockReturnValueOnce('something');
		const wrapper = shallow(
			<TreeNode {...props} getItemType={getItemType} branch={branch} leaf={leaf} />,
		);
		expect(leaf).toHaveBeenCalled();
		expect(wrapper.getElement()).toMatchSnapshot();
	});
});
