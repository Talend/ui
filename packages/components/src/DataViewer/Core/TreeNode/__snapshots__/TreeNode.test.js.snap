// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeNode should return a new branch 1`] = `
<div>
  new branch
</div>
`;

exports[`TreeNode should return a new leaf 1`] = `
<div>
  new leaf
</div>
`;
